### PR TITLE
refactor outbox/outbox entries to use single contract

### PR DESCRIPTION
**Pull request by DZGoldman**

### CODE REVIEW

File: packages/arb-bridge-eth/contracts/bridge/Outbox.sol [hunk 1]
```diff
@@ -178,18 +177,22 @@ contract Outbox is IOutbox, Cloneable {
...
-            outboxes[outboxIndex] = OutboxEntry(address(0));
+        require(!outboxEntry.spentOutput[uniqueKey], "ALREADY_SPENT");
+        require(calcRoot == outboxEntry.root, "BAD_ROOT");
+
+        outboxEntry.spentOutput[uniqueKey] = true;
+        outboxEntry.numRemaining--;
+
+        if (outboxEntry.numRemaining == 0) {
+            delete outboxEntries[outboxEntryIndex];
         }
```
Comment (hkalodner) [inline packages/arb-bridge-eth/contracts/bridge/Outbox.sol:196]: There shouldn't be any need to count number remaining or delete anything

**Review by hkalodner — CHANGES_REQUESTED** [review]

File: packages/arb-bridge-eth/contracts/bridge/Outbox.sol [hunk 2]
```diff
@@ -244,7 +241,11 @@ contract Outbox is IOutbox, Cloneable {
...
     }
 
-    function outboxesLength() public view returns (uint256) {
-        return outboxes.length;
+    function outboxEntriesLength() external view returns (uint256) {
+        return outboxEntries.length;
+    }
+
+    function outboxEntryExists(uint256 outboxEntryIndex) external view returns (bool) {
+        return outboxEntries[outboxEntryIndex].root != bytes32(0);
```
Comment (DZGoldman) [inline packages/arb-bridge-eth/contracts/bridge/Outbox.sol:249]: Q: is the assumption of this existence check sound? (or should outboxEntries just include an `exists` bool)

**Comment by DZGoldman** [comment]
TODO: add "outbox number" or "outbox version number" to each outbox

**Review by hkalodner — CHANGES_REQUESTED** [review]

File: packages/arb-bridge-eth/contracts/bridge/Outbox.sol (hunk 2 above)
Comment (DZGoldman) [reply packages/arb-bridge-eth/contracts/bridge/Outbox.sol:250]: it's really the equivalent usage on line 181 that matters (this one's just for client side convenience)

File: packages/arb-bridge-eth/contracts/bridge/Outbox.sol [hunk 3]
```diff
@@ -101,7 +100,7 @@ contract Outbox is IOutbox, Cloneable {
             uint256 numInBatch = data.toUint(33);
```
[the remark this answers is not in the document]
Comment (DZGoldman) [reply packages/arb-bridge-eth/contracts/bridge/Outbox.sol:101]: we use it in the OutboxEntryCreated event

**Review by hkalodner — APPROVED** [review]
LGTM

**Review by fredlacs — APPROVED** [review]
LGTM

**Pull request merged**